### DISCUSSION TITLE
Salvage value

### DISCUSSION THREAD

**Discussion by azev77** [infiniteopt/InfiniteOpt.jl#287] [Q&A] [answered]
A generic continuous time/non-stochastic optimal control problem is written:
![image](https://user-images.githubusercontent.com/7883904/188339196-0683e3f8-8728-4828-a21d-7c395071654a.png)
where:
F: cost/utility
S: salvage value/bequest value (only at the terminal value of the state variable)
f: dynamics/transition function
g: constraint functions

how can a user incorporate the Salvage function w/ your package?

**Comment by pulsipher** [comment, marked as answer]
Hi @azev77, I moved this over to the discussion forum since this is a usage question.

Yes, you can readily incorporate a terminal value using point variables. For instance:
```julia
using InfiniteOpt

model = InfiniteModel()
@infinite_parameter(model, t in [0, 1])
@variable(model, x, Infinite(t))

# Set objective where f(x(t)) = x(t)^2 and s(x(T)) = 2x(T)
@objective(model, Min, ∫(x^2, t) + 2 * x(1))
```

**Comment by azev77** [comment]
I just tried it and it works! 
I've been able to solve a large number of [Textbook problems](https://www.amazon.com/Optimal-Control-Theory-Applications-Management-dp-3030917444/dp/3030917444/ref=dp_ob_title_bk) w/ your package & compare the answer to the closed form solution.

I think it would help users a lot if a generic form of the optimal control problem was written down somewhere in the Readme or Docs & corresponding code was provided to show how to implement. 

Something like:
![image](https://user-images.githubusercontent.com/7883904/188508356-1c075b5d-bc3a-4586-8b74-4294338bbf88.png)
```julia
t0=0.0; tf=1.0;
u(s,c,t)   = -s;
S(s,t)     = 0.0; # s(s(T),T)
LOM1(s,c,t) = c;   # ṡ=LOM(s,c) 
g1(s,c,t)   =  c +1; # c +1>=0 or c>= -1
g2(s,c,t)   = -c +1; # -c +1>=0 
s_0=1.0;
#s_T 

opt = Ipopt.Optimizer   # desired solver
ns = 1_000;             # number of points in the time grid

m = InfiniteModel(opt)
@infinite_parameter(m, t ∈ [t0, tf], num_supports = ns)
@variable(m, s, Infinite(t)) ## state variables
@variable(m, c, Infinite(t)) ## control variables
@objective(m, Max, integral(u(s,c,t), t) + S(s(tf),tf))   #, weight_func = discount
@constraint(m, LOM_1, deriv(s, t) == LOM1(s,c,t))
@constraint(m, CON_1, g1(s,c,t) >= 0.0)                #@constraint(m, CON_1, 1.0 >= c >= -1.0) 
@constraint(m, CON_2, g2(s,c,t) >= 0.0) 
@constraint(m, s(0) == s_0)   ## initial condition

optimize!(m)
termination_status(m)
c_opt = value(c)
s_opt = value(s)
ts = supports(t)
opt_obj = objective_value(m) # V(B0, 0)
# using Plots
ix = 2:(length(ts)-1) # index for plotting
plot()
plot!(ts[ix],  s_opt[ix],  c=1, lab = "s")
plot!(ts[ix],  tt -> 1-tt, c=1, l=:dash, lw=3, lab = "s: closed form")
plot!(ts[ix],  c_opt[ix],  c=2, lab = "c")
plot!(ts[ix],  tt -> -1,   c=2, l=:dash, lw=3, lab = "c: closed form")
plot!([0.0],  seriestype = :hline, lab="", color="grey")
```

And make it very clear that this is for generic Optimal Control problems where the terminal time is fixed (not free)...

**Reply by pulsipher** [reply]
I'll look into providing some templates. The only difficulty is different disciplines use different general forms and nomenclature.

**Reply by azev77** [reply]
I moved this to a separate discussion: https://github.com/infiniteopt/InfiniteOpt.jl/discussions/291
